src/nodes.coffee

```diff
@@ -762,8 +748,43 @@ exports.Block = class Block extends Base
     return nodes[0] if nodes.length is 1 and nodes[0] instanceof Block
     new Block nodes
 
-  astProperties: ->
-    expressions: @expressions.map (child) => child.ast()
+  asExpressionStatementAst: (ast) ->
+    # return ast if ast.type in ['ImportDeclaration']
+
+    Object.assign
+      type: 'ExpressionStatement'
+      expression: ast
+    ,
+      extractAstLocationData ast
+
+  getExpressionAst: (expression) ->
+    ast = expression.ast()
+    # return ast if expression.isStatement o
+    @asExpressionStatementAst ast
+
+  bodyToAst: ->
+    @getExpressionAst(expression) for expression in @expressions
+
+  rootToAst: ->
+    programLocationData = @astLocationData()
+
+    programAst = Object.assign
+      type: 'Program'
+      sourceType: 'module'
```

Comment:
I’m on Node’s [modules](https://github.com/nodejs/modules) committee actually, so I can perhaps explain this. So JavaScript source files have a “parse goal,” which in a browser context is `script` or `module`, with `script` being traditional JavaScript and `module` being new ES6 stuff pulled in via `<script type="module"` or an `import` statement. Node doesn’t have `<script>` tags with a `type="module"` property, so some members of the Node community have been pushing the `.mjs` file extension to be Node’s version of `type="module"`, e.g. a place to store the metadata of a file’s parse goal outside of the source code itself.

So basically, ever since the introduction of ES modules now there’s a need to store at least this one piece of metadata about a file, its parse goal, outside of the file itself; either in `type="module"` or `.mjs`, or perhaps in `package.json` if I get my way. I’ve been arguing that `.mjs` isn’t very compatible with transpiled languages like CoffeeScript and TypeScript, since we would need to create our own new extension like `.mcoffee` to know when to output JavaScript files with an `.mjs` extension, and that causes a cascade of ecosystem concerns. We have the similar issue with knowing when to use `sourceType: 'module'`.

This is far from settled. Most of the browser community seems to be disinclined to start using `.mjs`, so there’s a split growing between browser-based ES modules that still use the `.js` extension and Node’s first attempt at ES modules support, the `--experimental-modules` implementation that’s shipped in the last few versions, that requires `.mjs`. Node still needs to somehow figure out how to support ES modules syntax in code input via `--eval`, which is how the `coffee` command executes code (the transpiled JS is executed via `node --eval`). The use case of extensionless files, like `/usr/local/bin/coffee` or `/usr/local/bin/babel`, may even cause Node to support a pragma like `"use module";`, the progeny of `"use strict";`. The spec settled on `type="module"` in the first place because people were trying to avoid a repeat of the annoyance of `"use strict";`, so it would be ironic if that’s where we ended up, but there’s a lot to be said for the parse goal living in the file itself somehow rather than outside of it in a file extension or nearby file or importing `<script>` tag. Technically Node can’t simply treat the presence of `import` or `export` as triggering module mode, as ES modules need not contain an `import` or `export` statement. (And before you suggest any more ways that this could perhaps be handled, search the [modules repo](https://github.com/nodejs/modules) because there’s a very good chance that anything you come up with has already been proposed. 😄)

So that’s a very long way of saying that I don’t know how we should decide this. We don’t need to be as strict as Node’s “cover all nonzero cases” mandate, so we could just look for an `import` or `export` as triggering `sourceType: 'module'` and call that good enough. But if any files within a project happen to lack such a statement, and it’s one random file within a whole project of module files, it probably _should_ be `sourceType: 'module'` too. So really the best approach might be to just leave out `sourceType` for now, assuming it’s not needed for Prettier support. Hopefully Node lands on the `package.json` approach, and CoffeeScript can copy whatever algorithm eventually Node settles on for determining whether a file in a package is an ES module, and once we start doing that we add in `sourceType`.

Reply:
Ok thanks for the interesting insights

> So really the best approach might be to just leave out `sourceType` for now

This is what I did for now. Does seem smart not to "commit" given some of what you mentioned

> assuming it’s not needed for Prettier support

It's not, but we should also consider ESLint to be a "first-class target" of this initial AST. During the last break I had (Sept) I made a lot of progress on the ESLint "custom parser"/plugin `eslint-plugin-coffee`. Currently (running against my branch as its `coffeescript` dependency) it has:
- full coverage of "recommended" rules included with ESLint
- full coverage of `eslint-plugin-react`
- nearly full coverage of _all_ rules included with ESLint
- nearly full coverage of what `coffeelint`/`coffeescope` covers

So it's about on par with where the Prettier plugin is (both maybe 80-90% of the way there)

So my plan from here is:
- continue prioritizing landing AST support here
- get both the Prettier and ESLint plugins to a point where they're ready to release by the time AST support lands/gets released

For me as a developer, Prettier is a very-nice-to-have but ESLint is crucial. And working on the ESLint integration has helped refine the AST implementation on my branch a lot (both in terms of accuracy and valuable things to expose)

So as far as `sourceType`, I've removed it for now but I want to:
- verify my assumption that it'll be easy to "monkeypatch" it back in as needed inside the ESLint plugin, since there are places there that expect it to be present
- consider this a TODO that can be informed both by the ongoing Node modules debate and by perhaps getting a little clearer sense of how it's used in ESLint rules